Updated documentation notes and example docstring style

davidorme opened the pull request:
# Description

This PR is a complete overhaul and extension of the pages describing the documentation setup, mostly off the back of an in person discussion we had about how to write and build documentation. There are no `vr` code changes, so I have again built the branch on RTD to make them easier to review. 

There is toy code module presenting a suggestion of the project style for docstrings, based on that discussion - if you have comments on the docstring style, you'll need to comment on the python file `docstring_style.py`.

There are then four HTML pages:

1. An [overview](https://virtual-rainforest.readthedocs.io/en/feature-doc_style/development/documentation/overview.html) of the docs tools and organisation, including the jupyter kernel thing that bit @alexdewar and @dalonsoa.
2. A sample python module showing the project docstring style is displayed as code in the [docstring style](https://virtual-rainforest.readthedocs.io/en/feature-doc_style/development/documentation/docstring_style.html) page.
3. That python code is then rendered, along with an explanation of how to get a module rendered in an [API generation](https://virtual-rainforest.readthedocs.io/en/feature-doc_style/development/documentation/api_generation.html).
4. Finally, I've updated the page on [Jupyter notebooks](https://virtual-rainforest.readthedocs.io/en/feature-doc_style/development/documentation/jupyter_notebooks.html) and the project format for them.

Fixes #144
Fixes #142 (by identifying the bug conditions and suggesting a different approach)

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Key checklist

- [x] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [x] All tests pass: `$ poetry run pytest`

jacobcook1995 commented on docs/source/development/documentation/overview.md:
````diff
@@ -0,0 +1,97 @@
+make html
+```
+
+The `make html` command will only builds pages for files that have changed recently, and
````
typo "will only builds pages"

jacobcook1995 commented on docs/source/development/documentation/api_generation.md:
```diff
@@ -0,0 +1,56 @@
+    jupytext_version: 1.13.8
+---
+
+# Rendered HTML for docstring example
```
Find the difference between this section title and it's name in the side bar "API Generation" a little bit jarring

jacobcook1995 replied:
Also as far as I can tell this page doesn't actually show the rendered output, which the link from `docstring_style.md` suggested it would

davidorme replied:
OK -  RTD runs `sphinx-build` from a different location. Just needed an alternative path in `conf.py` when running on RTD (or we could update where we build from...). All rendered now.

jacobcook1995 replied:
That renders now, but it seems to have broken the nesting of the API reference in the side bar

davidorme replied:
You mean the folding listings of content with files? So it has. 🤯 

I _much_ prefer the MkDocs explicit table of contents tree than `sphinx` ToC directives, which are just baffling.

jacobcook1995 commented on docs/source/development/documentation/api_generation.md:
`````diff
@@ -0,0 +1,56 @@
+Obviously, this explanatory box should be removed from real API pages.
+````
+
+```{eval-rst}
`````
Doesn't seem to be any output from this for some reason

jacobcook1995 commented on docs/source/development/documentation/api_generation.md:
```diff
@@ -0,0 +1,56 @@
+file in the `docs/source/api` folder. Typically, the file will start with the
+`jupytext` YAML metadata, then a markdown title and then an `automodule` instruction to
+insert all of the rest of the page content from the module docstrings. This text source
+of this page provides an example and can be seen using the 'View page source' link at 
```
I don't have a "View page source" link (as far as I can tell), just an "Edit on Github" link, are they the same thing?

davidorme replied:
Gah. Not quite, because you don't get the raw view of the contents, but that is one click away.

jacobcook1995 commented on docs/source/development/documentation/api_generation.md:
```diff
@@ -0,0 +1,56 @@
+
+The docstring content of a module is converted to HTML by including a small Markdown
+file in the `docs/source/api` folder. Typically, the file will start with the
+`jupytext` YAML metadata, then a markdown title and then an `automodule` instruction to
```
Feel like it might be worth spending a bit more time what each of these are for/why they are necessary

jacobcook1995 commented on docs/source/development/documentation/api_generation.md:
```diff
@@ -0,0 +1,56 @@
+docstring content and above the docstrings for those objects. The `members` option
+includes all public module members in the documentation. These options can be extended 
+to include other members if required. Here, the `special-members` option is used here
+to document the `__repr__` method. 
```
Is it the case that dunder methods have to be included in special methods or they are ignored?

davidorme replied:
It is. There is a similar option for `private-members`.

jacobcook1995 reviewed (changes requested):
Found most of this documentation easy to follow, apart from the API generation section, which I think could do with being a bit longer. The other problem there is that the docstring content doesn't seem to be actually getting rendered.

dalonsoa reviewed (approved):
Really nice! I've learnt a lot about integrating Jupyter stuff in the documentation. Documentation is rarely that didactic about the technical things!

davidorme commented:
@jacobcook1995 @dalonsoa I've pushed some commits to address Jacob's changes and a quick self-review after a re-read. I have also added some missing points:

* The QA steps run on documentation (in the Overview page).
* How to get your API page into the ToC (in the API Generation page).
* Update to the Jupyter notebook page - out of data information on QA tools at the bottom. Mostly moved into Overview and updated.

dalonsoa commented:
I'm not sure I'm seeing any of those changes in the RdD version, but it was good before, so probably so is now :)

davidorme commented:
Force reload maybe? My browser was definitely caching and the new content is there!

dalonsoa commented:
Now it is there! Cookies... always cookies. 

Anyway, nothing to add, really. It loos really nice and the explanations for developers about how things are arranged and how to setup their system is totally clear 👍

jacobcook1995 commented:
Yep all very clear to me as well!

vgro commented on docs/source/development/documentation/api_generation.md:
```diff
@@ -0,0 +1,74 @@
+from a [range of
+options](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-automodule)
+if required. Here, for example, the `special-members` option is used to make sure the
+`__repr__` and its docstrings are included method. This option covers all 'dunder'
```
'are included method'? is here a word missing?

vgro commented on docs/source/development/documentation/docstring_style.py:
```diff
@@ -0,0 +1,100 @@
+"""This is the documentation for the module. It does not start with a header line
+because a header is required at the top of the markdown source page where the API docs
+are will be inserted using the ``automodule`` declaration, so we do not repeat it here.
```
```suggestion
will be inserted using the ``automodule`` declaration, so we do not repeat it here.
```

TaranRallings reviewed (approved):
Looks great to me!

davidorme commented on docs/source/development/documentation/api_generation.md:
```diff
@@ -0,0 +1,74 @@
+  
+* A markdown header (`# A header`), which is shown as the page title. It is also used as
+  the text for links to this page from the table of contents, unless a shorter name is
+  provided in the `index.md` file. Typically, this header is the only markdown content
```
I've introduced `index.md` earlier and given the full path within `docs`.

davidorme commented:
Thanks all! @vgro - if you've got minor changes, you can make comments and suggestions and still approve the PR. The author can still update and resolve those comments, but doesn't then need to ask you for a re-review. If you request changes, you have to re-review it to accept those changes before it can be merged.

alexdewar commented on docs/source/development/documentation/overview.md:
````diff
@@ -0,0 +1,125 @@
+for writing Jupyter notebooks, making use of the [`jupytext`
+extension](https://jupytext.readthedocs.io/) to Jupyter.
+  
+```{admonition} More information
````
The "admonition" seems to imply that these blocks are for things you shouldn't be doing.

Is there a different block type for extra info? Or does the formatting look fine as is?

davidorme replied:
I completely agree that using `admonition` seems loaded, but that seems to be the generic name for this kind of box. There _are_ different kinds but they have fixed box titles and the generic `admonition` is the only one that allows you to set the title: https://jupyterbook.org/en/stable/reference/cheatsheet.html#admonitions

Weirdly, there doesn't seem to be an `info` or `information` option.

alexdewar replied:
Fair enough, just thought I'd check!

alexdewar reviewed (approved):
One query, but other than that it LGTM!

I didn't know about `jupytext`, but it looks cool.

vgro reviewed (approved):
Oh Ok, I didn't know that I could approve it with minor changes requested. here you go :-)

LivDaniel reviewed (approved):
Looks good. A bit over my head but it seems logical. My only thought was sometimes you give an explanation of what you don't use and why before saying what you should use and why, maybe start with what you recommend using and then say why the default option is no used after?
